src/pages/workspaces/workspace/_tests/WorkspaceMenu.test.js

```diff
@@ -0,0 +1,258 @@
+import '@testing-library/jest-dom'
+
+import { fireEvent, render, screen } from '@testing-library/react'
+import { useState } from 'react'
+import { div, h } from 'react-hyperscript-helpers'
+import { MenuTrigger } from 'src/components/PopupTrigger'
+import TooltipTrigger from 'src/components/TooltipTrigger'
+import { useWorkspaceDetails } from 'src/components/workspace-utils'
+import WorkspaceMenu from 'src/pages/workspaces/workspace/WorkspaceMenu'
+
+
+jest.mock('src/components/workspace-utils', () => {
+  const originalModule = jest.requireActual('src/components/workspace-utils')
+  return {
+    ...originalModule,
+    useWorkspaceDetails: jest.fn()
+  }
+})
+
+jest.mock('src/components/PopupTrigger', () => {
+  const originalModule = jest.requireActual('src/components/PopupTrigger')
+  return {
+    ...originalModule,
+    MenuTrigger: jest.fn()
+  }
+})
+
+jest.mock('src/components/TooltipTrigger', () => ({
+  ...jest.requireActual('src/components/TooltipTrigger'),
+  __esModule: true,
+  default: jest.fn()
+}))
+
+
+describe('WorkspaceMenu - undefined workspace', () => {
+  beforeEach(() => {
+    useWorkspaceDetails.mockReturnValue({ workspace: undefined })
+    MenuTrigger.mockImplementation(({ content }) => { return div([content]) })
+    TooltipTrigger.mockImplementation(({ content, children }) => {
+      const [open, setOpen] = useState(false)
+      return (div([
+        div(
+          {
+            onMouseEnter: () => {
+              setOpen(true)
+            },
+            onMouseLeave: () => {
+              setOpen(false)
+            }
+          },
+          [children]
+        ),
+        open && !!content && div([content])
+      ]))
+    })
+  })
+
+  it('renders Clone menu item as disabled', () => {
+    const onClone = jest.fn()
+    render(h(WorkspaceMenu, {
+      iconSize: 20, popupLocation: 'left',
+      callbacks: { onClone },
+      workspaceInfo: { name: 'example1', namespace: 'example-billing-project' }
+    }))
+    const menuText = screen.getByText('Clone')
+    expect(menuText).toHaveAttribute('disabled')
+  })
+
+  it('renders Clone tooltip', () => {
+    const onClone = jest.fn()
+    render(h(WorkspaceMenu, {
+      iconSize: 20, popupLocation: 'left',
+      callbacks: { onClone },
+      workspaceInfo: { name: 'example1', namespace: 'example-billing-project' }
+    }))
+    fireEvent.mouseOver(screen.getByText('Clone'))
+    expect(screen.queryByText('Cloning is not currently supported on Azure Workspaces')).toBeNull()
```

Comment:
I don't love this test-- even if a tooltip about Azure was incorrectly showing, the test would pass if the tooltip text was changed in the source code file. To make this test useful, I think a constant needs to be introduced for the text so it can be referenced in both places.

What I would prefer to see is that we are testing that the correct tooltip text **is** visible, but we don't show a tooltip during this initial loading period. Perhaps we should say something like "Workspace information is loading", but I think we would need to run that by UX.

Reply:
I also would be tempted fold the verification of the enable/disable state and the tooltip text into the same test case to share the setup code and make the relationship between them clearer. But I imagine some folks would prefer test cases to be as small and targeted as possible. @msilva-broad, what do you think?

Reply:
I think having seperate tests for the menu item checks and the tooltips checks are ok.  That gives better signal on what exactly broke if one of the tests fails.  It's ok however to group related assertions into each of those tests... like assert that all the menu items are rendered as expected.